File: src/com/medium/CombinationSum_39.java
/**
 * 
 */
package com.medium;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class CombinationSum_39 {
    // 2019年12月18日
    // Use DFS
    // Time Complexity: O(n^d)
    // Space Complexity:O(n)
    // n for candidate nums which less than target
    // d for depth that d = target/ min
    // Runtime: 5 ms, faster than 52.26%
    // Memory Usage: 37.6 MB, less than 100.00%
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> mResult = new ArrayList<>();
        Arrays.sort(candidates);
        dfs(candidates, target, 0, 0, new ArrayList<Integer>(), mResult);

        return mResult;
    }

    private void dfs(int[] candidates, int target, int sum, int mStart, List<Integer> mCur,
            List<List<Integer>> mResult) {
        if (sum > target) {
            return;
        }
        if (sum == target) {
            mResult.add(new ArrayList<Integer>(mCur));
            return;
        }

        for (int i = mStart; i < candidates.length; i++) {
            mCur.add(candidates[i]);
            dfs(candidates, target, sum + candidates[i], i, mCur, mResult);
            mCur.remove(mCur.size() - 1);
        }
    }
}

File: Golang_src/medium/combinationSum_39.go
package medium

// Use backtracking:
// S: all result of sum
// Time Complexity: O(S), upper bound: O(n*2^n)
// Space Complexity:O(target)
// Runtime: 4 ms, faster than 72.36%
// Memory Usage: 3.1 MB, less than 57.82%
func combinationSum(candidates []int, target int) [][]int {
	result := [][]int{}
	var dfs func(index int, tmp []int, sum int)

	dfs = func(index int, tmp []int, sum int) {
		if sum > target {
			return
		}
		if sum == target {
			newTmp := make([]int, len(tmp))
			copy(newTmp, tmp)
			result = append(result, newTmp)
			return
		}

		for i := index; i < len(candidates); i++ {
			tmp = append(tmp, candidates[i])
			// i for element can repeat, if change to i+1 cannot
			dfs(i, tmp, sum+candidates[i])
			// pruning
			tmp = tmp[:len(tmp)-1]
		}
	}
	// reset to previous node
	dfs(0, []int{}, 0)

	return result
}

// Use backtracking 2:
// S: all result of sum
// Time Complexity: O(S), upper bound: O(n*2^n)
// Space Complexity:O(target)
// Runtime: 11 ms, faster than 27.33%
// Memory Usage: 3 MB, less than 86.82%
func combinationSum2(candidates []int, target int) [][]int {
	result := [][]int{}
	keep := []int{}

	var dfs func(idx, cur int)
	dfs = func(idx, cur int) {
		if cur > target {
			return
		}
		if cur == target {
			tmp := make([]int, len(keep))
			copy(tmp, keep)
			result = append(result, tmp)
			return
		}

		for i := idx; i < len(candidates); i++ {
			keep = append(keep, candidates[i])
			dfs(i, cur+candidates[i])
			keep = keep[:len(keep)-1]
		}
	}

	dfs(0, 0)

	return result
}
